samu.c

```diff
@@ -79,17 +79,60 @@ warnflag(const char *flag)
 		fatal("unknown warning flag '%s'", flag);
 }
 
+enum {
+	maxArgs = 64
+};
+
+static int argc;
+static char *argv_combined[maxArgs];
+static char **argv = &argv_combined[0];
+
+void
+copycmdlineargs(int cmdline_argc, char *cmdline_argv[])
+{
+	int i = 1;
+	
+	while (argc < maxArgs -1 && i <= cmdline_argc) {
+		argv[argc] = cmdline_argv[i];
+		argc++;
+		i++;
+	}
+	argv[argc] = NULL;
+}
+
+void
+parseenvargs(char *s)
+{
+	char *p;
+		
+	p = strtok(s, " ");
+	while (p && argc < maxArgs - 1) {
+		argv[argc++] = p;
+		p = strtok(NULL, " ");
+	}
+	argv[argc] = NULL;
+}
+
 int
-main(int argc, char *argv[])
+main(int cmdline_argc, char *cmdline_argv[])
 {
 	char *builddir, *manifest = "build.ninja", *end, *arg;
 	const struct tool *tool = NULL;
 	struct node *n;
 	long num;
 	int tries;
-
-	argv0 = strrchr(argv[0], '/');
-	argv0 = argv0 ? argv0 + 1 : argv[0];
+	char *samuflags;
+
+	argv0 = strrchr(cmdline_argv[0], '/');
+	argv0 = argv0 ? argv0 + 1 : cmdline_argv[0];
+	argv[0] = cmdline_argv[0];
+	argc++; cmdline_argc--;
+	samuflags = getenv( "SAMUFLAGS" );
+	if (samuflags != NULL) {
+		samuflags = strdup(samuflags);
```

Comment:
Does this string need to be duplicated?

If it does, you need to check the result of `strdup`. Note also that `strdup` is not in C99 (though, I believe it will be added to C2X). I'm trying to limit the use of POSIX calls in case someone wants to pick up work on #12. To solve both of these problems, you could use `xmemdup` from `util.c`:

```
samuflags = xmemdup(samuflags, strlen(samuflags) + 1);
```

Reply:
strtok modifies the input string, hence I have to copy the string.

I made a xstrdup helper function using xmemdup function. It also handles the
case of passing a NULL to strdup, which by standard is undefined.

Currently I'm leaving a permanent leak of the copied environment, not sure, how
strict you are on leak checking with valgrind, otherwise I free it at the end, I think
it will solve itself anyway in the next step, when I split of the ARGV and the parseenv
parsing, so I leave it for now.

Though I believe a strdup also exists in the POSIXish parts of Windows
(https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strdup-wcsdup?view=vs-2019) it's best to avoid this function in this case.